### PR TITLE
Fix domain reload crash in inflate_info.

### PR DESCRIPTION
**Pull request by joncham**
The allocated memory is stored in a per image data structure (MonoImage::rgctx_template_hash) that can outlive domains. This results in a dangling pointer when the domain is reloaded.

- Should this pull request have release notes?
  - [X] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [X] Yes
  - [ ] No
- Do these changes need to be upstreamed to [mono/mono](https://github.com/mono/mono) or [dotnet/runtime](https://github.com/dotnet/runtime) repositories?
  - [ ] Yes
  - [X] No

Reviewers: please consider these questions as well! :heart:

**Release notes**

Fixed UUM-60632 @joncham:
Mono: Fix crash on domain reload, due to domain specific memory stored within image specific data structures.

**Backports**

2021.3, 2022.3, 2023.2, 2023.3

### CODE REVIEW

File: mono/mini/mini-generic-sharing.c [hunk 1]
```diff
@@ -726,17 +726,21 @@ inflate_info (MonoRuntimeGenericContextInfoTemplate *oti, MonoGenericContext *co
...
-		MonoClass *klass = mono_class_from_mono_type_internal (t);
+		MonoClass *inflated_klass = mono_class_from_mono_type_internal (t);
 		mono_metadata_free_type (t);
 
 		MonoMethod *method = mono_class_inflate_generic_method_checked (dele_info->method, context, error);
 		mono_error_assert_msg_ok (error, "Could not inflate generic method"); /* FIXME proper error handling */
 
-		// FIXME: Temporary
-		MonoDelegateClassMethodPair *res = (MonoDelegateClassMethodPair *)mono_domain_alloc0 (domain, sizeof (MonoDelegateClassMethodPair));
+		MonoDelegateClassMethodPair* res = NULL;
```
Comment (scott-ferguson-unity) [inline mono/mini/mini-generic-sharing.c:735]: Does a similar change need to be made at https://github.com/Unity-Technologies/mono/blob/e492b1480b46367e556e2a092fe5c026492b16d9/mono/mini/mini-generic-sharing.c#L708?  Or this that not reproducible?

Comment (joncham) [reply mono/mini/mini-generic-sharing.c:735]: In theory yes, but I didn't hit that code path at all. If we can figure out how to exercise that path I am fine fixing it as well.

There are also others that are only triggered when running with value type generic sharing (gsharedvt), which is not enabled in our fork (and never was).

**Review by scott-ferguson-unity — APPROVED** [review]

**Review by UnityAlex — APPROVED** [review]

File: mono/mini/mini-generic-sharing.c [hunk 2]
```diff
@@ -763,6 +767,9 @@ free_inflated_info (MonoRgctxInfoType info_type, gpointer info)
 	case MONO_RGCTX_INFO_CAST_CACHE:
 		mono_metadata_free_type ((MonoType *)info);
 		break;
+	case MONO_RGCTX_INFO_DELEGATE_TRAMP_INFO:
+		g_free (info);
```
Comment (bholmes) [inline mono/mini/mini-generic-sharing.c:771]: Is this alway the right free?  Can this pointer come from the `mono_image_alloc0` aka `temporary == false` case?

Comment (joncham) [reply mono/mini/mini-generic-sharing.c:771]: The convention is that this function `free_inflated_info` is only called for `temporary == true` cases. So this should be correct if there is not a bug in any calling code.

**Review by bholmes — APPROVED** [review]

**Pull request merged**